common/middleware/src/line_rules.rs

```diff
@@ -0,0 +1,293 @@
+use crate::{Middleware, Status};
+use http::types::body::LineBufferMut;
+use regex::bytes::Regex;
+use std::cmp;
+use thiserror::Error;
+
+static REDACT_BYTES: &[u8] = "[REDACTED]".as_bytes();
+
+pub struct LineRules {
+    exclusion: Vec<Regex>,
+    inclusion: Vec<Regex>,
+    redact: Vec<Regex>,
+}
+
+#[derive(Clone, Debug, Error)]
+pub enum LineRulesError {
+    #[error(transparent)]
+    RegexError(regex::Error),
+}
+
+impl LineRules {
+    pub fn new(
+        exclusion: &[String],
+        inclusion: &[String],
+        redact: &[String],
+    ) -> Result<LineRules, LineRulesError> {
+        Ok(LineRules {
+            exclusion: map_to_regex(exclusion)?,
+            inclusion: map_to_regex(inclusion)?,
+            redact: map_to_regex(redact)?,
+        })
+    }
+
+    /// Applies inclusion and exclusion rules and replaces the redacted values.
+    fn process_line<'a>(
+        &self,
+        line: &'a mut dyn LineBufferMut,
+    ) -> Status<&'a mut dyn LineBufferMut> {
+        let value = line.get_line_buffer().unwrap();
+
+        // If it doesn't match any inclusion rule -> skip
+        if !self.inclusion.is_empty() && !self.inclusion.iter().any(|r| r.is_match(value)) {
+            return Status::Skip;
+        }
+
+        // If any exclusion rule matches -> skip
+        if self.exclusion.iter().any(|r| r.is_match(value)) {
+            return Status::Skip;
+        }
+
+        if !self.redact.is_empty() {
+            return self.redact(value.to_owned(), line);
+        }
+
+        Status::Ok(line)
+    }
+
+    fn redact<'a>(
+        &self,
+        value: Vec<u8>,
+        line: &'a mut dyn LineBufferMut,
+    ) -> Status<&'a mut dyn LineBufferMut> {
+        let mut matches: Vec<(usize, usize)> = vec![];
```

Comment:
This might be a good place to use SmallVec? It depends how ofter we expect to find something to redact.

If there's no match 99% of the time then Vec makes sense as it doesn't allocate unless there's a match

If there are frequently 1-~5 or so matches then SmallVec would save a lot of allocations

If there are frequently more than 4 or 5 matches then it's porbably best for it to be a Vec again.


I think Vec is probably right for now, but I'm interested to hear your thoughts.

Reply:
From the use case, I think inclusion/exclusion rules match often and redact rules match rarely.